### ISSUE TITLE
Match routes by method, allow combinators to return errors

### ISSUE THREAD

**Issue by seanhess** [haskell-servant/servant#410]
Hi, I have run into an issue in 0.4. I hear that you've re-done the routing in 0.5, and I'm wondering if this is addressed. 

Routes are usually differentiated by the request method, but the request method isn't specified until the very last combinator. See below:

```
type UserAPI
   "users" :> Get '[JSON] [User]
   "users" :> ReqBody '[JSON] User :> Post '[JSON] User
   "users" :> Capture "userID" Text :> Get '[JSON] User
```

I think of this as follows:

    GET /users
    POST /users
    GET /users/:userId

But servant always uses earlier combinators to see if a route matches, so when it chooses between these two routes, it's *really* matching it this way:

    /users, method == GET
    /users, has request body, method == POST
    /users, has capture "userId", method == GET

This is a problem when I want to write a combinator that sends back errors when the request is invalid, like during authentication, or multipart uploads. See http://stackoverflow.com/questions/36025826/servant-combinator-doesnt-fall-through-to-subsequent-alternatives for a good description of the problem. 

I've created a combinator that works around this: https://gist.github.com/seanhess/1aa69806430a312936e8

```
type UserAPI
   "users" :> Get '[JSON] [User]
   "users" :> Method POST :> MultipartUpload :> Post '[JSON] 
   "users" :> Capture "userID" Text :> Get '[JSON] User
User
```

Basically I feel like we should be able to differentiate routes by the method early, and allow other combinators to return errors rather than just saying the route doesn't match. 

Is this addressed in 0.5? How? If not, how would you solve this problem? I've also run into it when creating authentication combinators, like `HasAuth Admin`.

**Comment by jkarni** [comment]
This is in fact addressed in 0.5. Combinators now add continuations to a datatype that knows in what order to process them. In the `basic auth` [PR](https://github.com/haskell-servant/servant/pull/375/files#diff-350ecd5a117c9bc3dd30e3b9feceb5ceR468), for example, `addAuthCheck` is a function that adds the continuation in the 'right place' (i.e. corresponding to the order of processing of HTTP data that is sensible). So now, independently of where combinators appear in the route API, they get processed as:

- Path 
- Method
- Body
- etc.

This [module](https://github.com/haskell-servant/servant/blob/master/servant-server/src/Servant/Server/Internal/RoutingApplication.hs) has the details. Note in particular the `Delayed` datatype and it's functions.

**Comment by jkarni** [comment]
Closing, but feel free to re-open or ask further questions!

**Comment by seanhess** [comment]
Hooray! That's perfect, and exactly how I would expect it to work. Now I just need to get on to 0.5 soon. Thanks @jkarni.
